src/System.Diagnostics.DiagnosticSource/src/HttpClientServerExample.cs

```diff
@@ -0,0 +1,220 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+namespace AspNetCore.Hosting
+{
+    /// <summary>
+    /// Example of Http server instrumentation.
+    /// </summary>
+    class AspNetCoreExample
+    {
+        public AspNetCoreExample(ITextPropagationFormat customPropagationFormat)
+        {
+            // get customPropagationFormat from DI
+            if (customPropagationFormat != null)
+            {
+                // set it for HttpClient
+                DiagnosticSource.HttpPropagationFormat = customPropagationFormat;
+            }
+
+            this.customPropagationFormat = customPropagationFormat;
+        }
+
+        private static readonly DiagnosticListener MySource = new DiagnosticListener("HttpInExample");
+
+        // Get implementation from DI
+        private readonly ITextPropagationFormat customPropagationFormat;
+
+        private void HttpIn(HttpRequest request)
+        {
+            bool hasListener = MySource.IsEnabled();
+
+            Activity ActivityFactory()
+            {
+                Activity activity = new Activity("httpin");
+
+                // if no custom propagation is defined - set activity.traceparent from header
+                if (customPropagationFormat == null)
+                {
+                    if (request.Headers.TryGetValue("traceparent", out var traceparent))
+                    {
+                        activity.W3CId = traceparent;
+                        if (request.Headers.TryGetValue("tracestate", out var tracestate))
+                        {
+                            activity.Tracestate = tracestate;
+                        }
+                    }
+                }
+                else // otherwise use custom format
+                {
+                    customPropagationFormat.Extract(request.Headers, (headers, s) =>
+                    {
+                        headers.TryGetValue(s, out string value);
+                        return value;
+                    }, activity);
+                }
+
+                return activity;
+            }
+
+            var actualActivity = MySource.StartActivityIfEnabled(ActivityFactory, request, null, () => new { /*httpcontext*/});
```

Comment:
Why do we bother with the Func<T> and StartActivityIfEnabled.  It does not seem to create an interesting abstraction but does make the code less efficient (significantly so, closure capture (which is what you are doing to get 'request')  is an object allocation which we DON'T want in the non-sampled case.

Reply:
Ignore my comment for now.   I see real code below that uses StartActivityIfEnabled below that may answer my question.